### ISSUE TITLE
Testing and validating the l1norm and peak counts for both low and high resolution simulation

### ISSUE THREAD

**Issue by dlanzieri** [LSSTDESC/DifferentiableHOS#10]
This issue is to asses the impact of low resolution N-body simulation (boxsize 64 Mpc/h ) on the weak lensing statistic and to track the development needed to validate the Peak counts and the l1norm. 

**Goal:**

1. Decide if the low-resolution simulations can be used for our purpose 
2. Validate the l1-norm and peak counts within the limit of baryonic effects and Born approximation.

These goals incorporate several smaller issues.

**Task:**

- [ ] : Compare the angular power spectrum, the l1norm and the peak counts  averaged over 100 maps realized by simulating a volume of 64  and 128 Mpc/h.
- [ ] :  Understand why the peak counts from our maps are higher than the peak counts from the  kTNG maps
- [ ] : Identify  the optimal k-range and bin setting for the peak counts
- [ ] : Increase the number of bins and scale in the l1norm computation 
- [ ] : Identify the optimal smoothing to avoid artifact in the final maps
- [ ] : Compare and validate our results with some references taking into account the systematic due to the Baryonic effects
- [ ] : Compare and validate our results with some references taking into account the limits of the Born approximation

**Comment by dlanzieri** [comment]
We obtained the optimal parameters for the PGD by implementing this [script](https://github.com/LSSTDESC/DifferentiableHOS/blob/026c5d2bc46238801dedc085f9b04bf95acf21e0/scripts/fit_pgd.py).
It's just a rearranged version of the [script](https://github.com/DifferentiableUniverseInitiative/flowpm/blob/6c45b8b97b8bbbc99b833b1aa2e46a18632cdcd5/scripts/fit_pgd.py) described in this [issue](https://github.com/DifferentiableUniverseInitiative/flowpm/pull/90) adapted to the lightcone construction.
I used these parameters to generate 1000 maps.   
Then I computed the angular power spectrum for each map and I averaged over all the 1000 samples.
Here the angular power spectrum compared to the _halofit_ predictions:
![cldll_clktng_64 (1)](https://user-images.githubusercontent.com/72620117/132030162-3896df62-22e6-41c1-8fd3-eef38c36bf04.png)

I started to  compare our results to the _K-TNG_ only Dark simulations. 
This is how two general maps from the simulations look like:
![maps_comp](https://user-images.githubusercontent.com/72620117/132030970-94712f38-94a9-4689-a978-bed4bfd5f130.png)
and if we see them with the same color-scale:
![maps_comp_same](https://user-images.githubusercontent.com/72620117/132031112-c8bf17b2-e8a5-4227-8be2-30a864c1962d.png)
and after computing also the averaged power spectrum over 1000 _K-TNG_ maps, I compared the two power spectra:

![cldll_clktng_64](https://user-images.githubusercontent.com/72620117/132031499-545cd05c-0818-4f81-9e47-ee767c8b9f1d.png)

@EiffL what do you think about? Does it look reasonable? 
N.B. All the results are without smoothing and noise.

**Comment by EiffL** [comment]
arf >.< that map looks kind of horrible

**Comment by EiffL** [comment]
Ok, @dlanzieri @modichirag  I got some good news. I've investigated this a bit more and came to several conclusions:

## Offline vs Online PGD
I've implemented PGD here with cic correction, and simply using (1 - pk/pkref)**2 as the loss function, no weights. The boxes here are 200^3 Mpc/h  and 128^3, B=1.

- **Offline PGD fails because we are asking it to correct for too big deviations at once**, and so it creates some weird artifacts. This can be seen at the level of the lensplanes a little bit, for instance here a lensplane at z=0 with PGD turned off (left)  and turned on (right):
![image](https://user-images.githubusercontent.com/861591/132125641-44bd1017-8f42-4f35-a95e-6ad35429b38d.png)
PGD creates some very sharp clusters and actually changes the shape of some of the structures, I think this implies that the displacements are quite large. And we can also look at the effect on mass maps, here is for instance a mass map with PGD turned off (left)  and turned on (right):
![image](https://user-images.githubusercontent.com/861591/132125503-0a816c16-e07a-47c1-a8c2-3cad63dade51.png)
Filamentary structures are overly-amplified. 

- **Online PGD works nicely** because at each timestep it only needs to perform a small correction to the displacement. 
Same comparison of lensplanes:
![image](https://user-images.githubusercontent.com/861591/132125700-aea091a7-0323-40fa-982a-6344d3005774.png)
Here, to me at least, it looks like the structures are not modified but simply sharpened. Here is the result on a mass map  with PGD turned off (left)  and turned on (right):
![image](https://user-images.githubusercontent.com/861591/132125599-944cb958-ceda-4f30-914d-e895f399218e.png)
This looks much nicer, and we don't see the weird filamentary structures.

**=> So we should use online PGD**

## Impact of simulation volume

So, this is something that Joachim warned us about, the simulation volume needs to be large enough otherwise the large scales slip away during evolution. This matters for us when we fit PGD, we don't want PGD to be fixing the large scale stuff, so we would prefer the simulation to be already well behaved there.
Here is a small illustration off the problem, on a box of volume 64 Mpc/h ^3
![image](https://user-images.githubusercontent.com/861591/132126013-b581e150-21da-4cd3-a0f1-0c18d6e1be6c.png)
The shaded lines are different realisations of the simulation down to z=0, what we see is that on the largest scale, we are systematically biased below the expected power spectrum. If we believe Joachim it's because we don't have enough linear scales to anchor us.
If I run a simulation on a box of volume 256 Mpc/h^3 I get this:
![image](https://user-images.githubusercontent.com/861591/132126081-d3fc8f41-d980-4bce-a152-923baec712bf.png)
Obviously this is not doing a great job on non-linear scales, but the large scales are nicely unbiased.

**=> So, if we now can use PGD to fix small scales, we should prefer larger volumes, to make sure the linear scales remain ok in the simulation. A simulation volume of 205 Mpc/h^3 would be the same at TNG 300 and seems like a good choice**

## Memory cost of Jacobians

The main reason why we have been confined to using 64^3 simulations is that we were running out of RAM even on full Haswell nodes when trying to get the gradients of a simulation. For reference, I can run 128^3 sims on my GPU, which has 12 GB of RAM, so with gradients, we should be looking at something like ~30 GB and this should fit on a Haswell node without problems. It should even run on an A100 GPU for whenever we finally get Perlmutter access.

I investigated this a little bit more, and realized that by default tape.jacobian() is trying to run several gradient evaluation in parallel, to speed up computations. So this blows up memory consumption and easily blows over 128GB, crashing the job. But this can be fixed by using the following arguments of the jacobian function:
```python
tape.jacobian(... , experimental_use_pfor=False, parallel_iterations=1)
```
This constrains the gradient evaluation to a single dimension of the output stat at a time, and so memory consumption stays at ~2x the cost of the forward pass. Compute time will scale linearly with the dimension of the output but that's ok.

I ran a few comparisons to see what was the tradeoff in memory resolution on my local machine (execution might differ slightly on cori):

**64^3,  nlens = 22 (boxsize=100Mpc)**:
- ~30 GB of ram 

**100^3,  nlens = 11 (boxsize=200Mpc)**:
- ~30 GB of ram 
- 10 mins per grad evaluation for output of size 7

**128^3,  nlens = 11 (boxsize=200Mpc)**:
- ~50 GB of ram 
- 4.5 mins per grad evaluation for output of size 7

So, interestingly, by using a larger volume we can offset the cost of higher number of particules because we need fewer steps. 

**=> To me, this would point to trying to use 128^3 with boxsize of 200Mpc, assuming we can PGDize it well enough to not need any increase in resolution** This would let us use 2 processes in parallel on a Haswell node, and output about 20 jacobians per hour per node.  


## Action items

Here I will just summarize the things we need now to get a script that can compute maps and  Jacobians at the desired resolution:

- Online PGD function in FlowPM
- Online PGD training script tuned to our particular settings here
- Updated maps/jacobian script in DifferentiableHOS

I'll open dedicated issues for each of these items as I go through them.

**Comment by EiffL** [comment]
I added a script for training PGD parameters + added a script to generate maps in `u/EiffL/update_settings`. I've defaulted to using a boxsize of 205 Mpc/h^3, 128^3 particles, 11 nlens, and online PGD. 

To train pgd params:
```bash 
$ python scripts/fit_online_pgd.py
```
Then you can use these PGD params in other scripts, see for instance this [notebook](https://github.com/LSSTDESC/DifferentiableHOS/blob/u/EiffL/update_settings/notebooks/Prototype_SRD_Y1_online_pgd.ipynb), or this new [compute_maps.py](https://github.com/LSSTDESC/DifferentiableHOS/blob/u/EiffL/update_settings/scripts/compute_maps.py).

Here are a few promising plots from the notebook:
![image](https://user-images.githubusercontent.com/861591/132144188-42e0b1b9-7d3f-4b29-a774-3e1cedf8e9da.png)
![image](https://user-images.githubusercontent.com/861591/132144190-90702093-f785-485e-ae2b-dc66549b0bde.png)
These two plots above are without smoothing, then with one arcmin smoothing:
![image](https://user-images.githubusercontent.com/861591/132144209-65ff7cd4-1b1a-475b-88e2-4c07aa666062.png)
![image](https://user-images.githubusercontent.com/861591/132144212-256bec2b-680d-48a5-8902-faa325473890.png)
![image](https://user-images.githubusercontent.com/861591/132144214-6f560c43-1240-4d0a-8030-539beb901b1e.png)

**Comment by EiffL** [comment]
I've opened a PR #11 but it's a little bit of a mess, we should try to remove any duplication of the code we use to generate a convergence map,  so move all of that to a module file under `DifferentiableHOS`, which can then be reused from multiple scripts.

**Comment by dlanzieri** [comment]
Hi @EiffL ! I generated severals maps by using the  PGD parameters we got as you described above. I also computed the covariance matrix and the Fisher matrix. May you have a look at this [notebook](https://github.com/LSSTDESC/DifferentiableHOS/blob/751ed6982fbbb688a119dec3b3fef1827fb72ae8/notebooks/Prototype_SRD_Y1_load_results.ipynb) and let me know what you think?

**Comment by EiffL** [comment]
It looks good @dlanzieri! Do you have everything you need to do the same for wavelet peak counts?

**Comment by EiffL** [comment]
Any luck with the peak counts?

**Comment by dlanzieri** [comment]
@EiffL I got the first Jacobian, but I want to investigate more the optimal k-range and bin setting for the peak counts. 
![peakcount](https://user-images.githubusercontent.com/72620117/133750277-7cb7d48b-8215-4673-9703-b60f7b89f361.png)
I don't  know if this result comes from the noise of a single map or there is something wrong. I'll investigate more. 
Meanwhile I started to compare the Jacobian of the power spectrum obtained from the maps without noise and smoothing to the Jacobian of the reference power spectrum. I remember that in our previous notebook we showed the normalized jacobians, computed by dividing the jacobians by the theory power spectrum. In this [notebook](https://github.com/LSSTDESC/DifferentiableHOS/blob/30e6b412162cc4077b158d546c1f4ed63195543c/notebooks/Prototype_SRD_Y1_halofit_comp.ipynb) I tried to reproduce the same plots.  The results don't seem so good, at least for the Normalized Jacobian.

**Comment by EiffL** [comment]
oh nice! why do you say this looks weird?

**Comment by EiffL** [comment]
we didnt normlize by halofit, but by the the corresponding power spectrum

**Comment by dlanzieri** [comment]
@EiffL not weird, I juste want to say that we don’t have a benchmark,  so I want to understand if this trend comes from the fact I used a single map (so if now it's just a case), from the bin setting I used or it's as indeed it really ought to be. I have no idea of how  it really ought to be.

**Comment by EiffL** [comment]
So also, remember @dlanzieri that you need to rebin the theory power spectrum in the same way as you rebin the measured PS.
And can you add a plot of your mean measured power spectrum vs theory?

**Comment by EiffL** [comment]
So, hum.... these plots for the power spectrum as worrying....

**Comment by dlanzieri** [comment]
I did it to compute the theory power spectrum (I don't know why some output in the notebook have been deleted), but I don't know if I did it in the right way to compute the theoretical Jacobian. 
I mean, I computed  the theory power spectrum  using the original  ell range and then I computed the rebin in the same way of the measured power spectrum. But for the Jacobian I used the "rebinned"   ell (that is what I did also for the measured ps)

**Comment by EiffL** [comment]
in your `mean_fn` you need to sample the PS at the same points as what you get from the TF measurement code and rebin the PS, inside that function.

**Comment by dlanzieri** [comment]
ok, perfect! Something like this:
```ruby
def mean_fn(p):
  cosmo = jc.Planck15(Omega_c=p[0], sigma8=p[1])
  m = jc.angular_cl.angular_cl(cosmo, l.numpy(), probes)
  m = m[2:46]
  m = rebin(m, 11)
  return m.flatten()
jac_mean = jax.jit(jax.jacfwd(mean_fn))
params = np.array([cosmo.Omega_c, cosmo.sigma8]) 
dmu = jac_mean(params)
```
?

**Comment by EiffL** [comment]
:-) you'll just need to rewrite the rebin function in JAX

**Comment by dlanzieri** [comment]
Hi @EiffL, I updated this [notebook](https://github.com/LSSTDESC/DifferentiableHOS/blob/1fb695974576dab3a37416e65b3a91415f034663/notebooks/Prototype_SRD_Y1_load_results.ipynb) :)
This is the Fisher matrix with our current implementation :
![Fisher (1)](https://user-images.githubusercontent.com/72620117/133888504-9151873c-fe9a-4148-b0b3-80df58d4acc0.png)
)

**Comment by modichirag** [comment]
I just wanted to say that this looks great, right? :)

**Comment by EiffL** [comment]
Ohh nice! Yeah that's what we are talking about :-)
